src/core/lib/transport/promise_endpoint.h

```diff
@@ -0,0 +1,315 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef GRPC_CORE_LIB_TRANSPORT_PROMISE_ENDPOINT_H
+#define GRPC_CORE_LIB_TRANSPORT_PROMISE_ENDPOINT_H
+
+#include <grpc/support/port_platform.h>
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <functional>
+#include <memory>
+#include <utility>
+
+#include "absl/base/thread_annotations.h"
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+#include "absl/types/optional.h"
+
+#include <grpc/event_engine/event_engine.h>
+#include <grpc/event_engine/slice.h>
+#include <grpc/event_engine/slice_buffer.h>
+#include <grpc/slice_buffer.h>
+#include <grpc/support/log.h>
+
+#include "src/core/lib/gprpp/sync.h"
+#include "src/core/lib/promise/activity.h"
+#include "src/core/lib/promise/poll.h"
+#include "src/core/lib/slice/slice.h"
+#include "src/core/lib/slice/slice_buffer.h"
+
+namespace grpc {
+
+namespace internal {
+
+class PromiseEndpoint {
+ public:
+  PromiseEndpoint(
+      std::unique_ptr<grpc_event_engine::experimental::EventEngine::Endpoint>
+          endpoint,
+      grpc_core::SliceBuffer already_received);
+  ~PromiseEndpoint();
+
+  auto Write(grpc_core::SliceBuffer data) {
+    grpc_core::MutexLock lock(&write_mutex_);
+
+    /// Previous write result has not been polled.
+    GPR_ASSERT(!write_result_.has_value());
+
+    /// TODO: Is there a better way to convert?
+    grpc_slice_buffer_swap(data.c_slice_buffer(),
+                           write_buffer_.c_slice_buffer());
+
+    const std::function<void(absl::Status)> write_callback =
+        [this](absl::Status status) {
+          grpc_core::MutexLock lock(&write_mutex_);
+          write_result_ = status;
+          write_waker_.Wakeup();
+        };
+    endpoint_->Write(write_callback, &write_buffer_,
+                     nullptr /* uses default arguments */);
+
+    return [this]() -> grpc_core::Poll<absl::Status> {
+      grpc_core::MutexLock lock(&write_mutex_);
+      if (!write_result_.has_value()) {
+        write_waker_ = grpc_core::Activity::current()->MakeNonOwningWaker();
+        return grpc_core::Pending();
+      } else {
+        const auto ret = *write_result_;
+        write_result_.reset();
+        return ret;
+      }
+    };
+  }
+
+  auto Read(size_t num_bytes) {
+    grpc_core::MutexLock lock(&read_mutex_);
+
+    /// Previous read result has not been polled.
+    GPR_ASSERT(!read_result_.has_value());
+
+    /// Should not have pending reads
+    GPR_ASSERT(pending_read_buffer_.Count() == 0u);
+
+    if (read_buffer_.Length() < num_bytes) {
+      const std::function<void(absl::Status)> read_callback =
```

Comment:
const auto (no need for type erasure here)

also... consider moving `num_bytes`, `read_callback` into members on the `PromiseBasedEndpoint`: storing more than one pointer on a callback will force a memory allocation when we call `Endpoint::Read` and convert this callback into an `AnyInvocable`.